Constrained LQR controller

shrenikm opened the pull request:
Implementation of a Constrained LQR controller and other small fixes.

## Controller Class

The controller is implemented through the `ConstrainedLQRController` class which is a subclass of `LeafSystem`. Although the controller implementation is affine, it does not extend `AffineController` due to difficulties in incorporating the required ports.

`constrained_lqr_controller_test.cc` is the unit test script for the controller class.

`run_constrained_lqr_controller.cc` is a run script that implements the controller for the fixed and floating base Cassie versions. The base type is given through the gflag `floating_base`. Controller gains have been tuned for the floating base implementation.

## RBT Visualizer

`rbt_visualizer.cc` has been added for floating base simulation (Which is not currently supported by `MultibodyPlant`). The corresponding entries have also been added to the procman script.

## Other fixes

* Fixed a bug in `AffineController` (As pointed out by @Nanda-Kishore-V ) in one of the setter methods. Added additional tests to the `AffineController` test script to account for this and a few other methods that weren't tested.
* Fixed a bug in `FixedPointSolver` where `AddSpreadingNormalForces` did not add the expression constraint after constructing the expression.
* Changed incorrect namespaces in some of the previous test files in `multibody_sovlers` and `contact_toolkit`.

mposa commented on examples/Cassie/run_constrained_lqr_controller_sim.cc:
```diff
@@ -0,0 +1,472 @@
+ * It serves as an interfacing connector between the RigidBodyPlant's state
+ * output port and ConstrainedLQRController's info input port.
+ */
+class StateConnector : public LeafSystem<double> {
```
It feels like this should be moved to `multibody` and given a more appropriate name. It's acting like a state estimator from a simulated plant, but not currently copying over the efforts. I think it's OK, for now, to avoid adding in the effort logic (which will require an additional input port), but others would have use for a block like this.

mposa commented on examples/Cassie/run_constrained_lqr_controller_sim.cc:
```diff
@@ -0,0 +1,472 @@
+  vector<string> cost_ignore_position;
+  if (FLAGS_floating_base) {
+    cost_ignore_position = {"base_x",    "base_y",     "base_z",
+                            "base_roll", "base_pitch", "base_yaw"};
```
Why ignore angular terms?

mposa commented on systems/controllers/affine_controller.cc:
```diff
@@ -7,38 +7,40 @@ using Eigen::VectorXd;
-  auto K = params->get_K();
-  auto desired_state = params->get_desired_state();
-  auto E = params->get_E();
+  MatrixXd K = params->get_K();
```
Using `MatrixXd` here, I believe, forces a copy. `params` returns a `const Eigen::Map<Eigen::MatrixXd>`, which is a reference to the data (rather than a copy).

mposa commented on systems/controllers/constrained_lqr_controller.h:
```diff
@@ -0,0 +1,121 @@
+ * ConstrainedLQRController class that implements an LQR controller that also
+ * takes into account constraints in the state space.
+ * The controller follows an affine structure. It is not a subclass of
+ * AffineControlelr due to difficulties in incorporating the required input
```
nit-typo

mposa commented on systems/controllers/constrained_lqr_controller.h:
```diff
@@ -0,0 +1,121 @@
+   * The constructor computes K and E and is the major computational block of
+   * the controller class.
+   */
+  ConstrainedLQRController(const RigidBodyTree<double>& tree, Eigen::VectorXd q,
```
All of these inputs should probably be const references.

mposa commented on systems/controllers/constrained_lqr_controller.cc:
```diff
@@ -0,0 +1,141 @@
+    MatrixXd Q, MatrixXd R, ContactInfo contact_info)
+    : tree_(tree),
+      contact_info_(contact_info),
+      contact_toolkit_(
```
Question: why a unique_ptr?

shrenikm replied:
I need `tree` and `contact_info` to create the object (It doesn't have a default constructor). So I couldn't really define it as a private non-pointer variable.

mposa commented on systems/controllers/constrained_lqr_controller.cc:
```diff
@@ -0,0 +1,141 @@
+  // Input port that takes in an OutputVector containing the current Cassie
+  // state
+  input_port_info_index_ =
+      this
```
Linebreak formatting looks off here

shrenikm replied:
For some reason my linting formats it this way. I'm using the Cland formatter.

mposa commented on systems/controllers/constrained_lqr_controller.cc:
```diff
@@ -0,0 +1,141 @@
+
+  // Output port that outputs the efforts
+  output_port_efforts_index_ =
+      this->DeclareVectorOutputPort(TimestampedVector<double>(num_efforts_),
```
Is this better than 
```
output_port_efforts_index_ =
    this->DeclareVectorOutputPort(TimestampedVector<double>(num_efforts_),
        &ConstrainedLQRController::CalcControl).get_index();
```
?

mposa commented on systems/controllers/constrained_lqr_controller.cc:
```diff
@@ -0,0 +1,141 @@
+  MatrixXd J(J_tree.rows() + J_contact.rows(), num_positions_);
+  J << J_tree, J_contact;
+
+  // Computing F
```
Add brief explanation of what `F` is

mposa commented on systems/controllers/constrained_lqr_controller.h:
```diff
@@ -0,0 +1,121 @@
+namespace systems {
+
+/*
+ * ConstrainedLQRController class that implements an LQR controller that also
```
Somewhere in this description would be a good place to reference the appropriate publication.

mposa commented on systems/controllers/constrained_lqr_controller.cc:
```diff
@@ -0,0 +1,141 @@
+  DRAKE_DEMAND(Q_.rows() == Q_.cols());
+  DRAKE_DEMAND(R_.rows() == R_.cols());
+  DRAKE_DEMAND(A_.rows() == Q_.rows());
+  DRAKE_DEMAND(B_.cols() == R_.rows());
```
Aren't all of these checks guaranteed by the linear algebra above? I think the only one that isn't guaranteed is the last `B.cols == R.rows`. It's probably better to check that `Q` and `R` inputs have appropriate size, then assume that the matrix multiplies result in proper dimensions.

mposa commented:
Is this waiting for a re-review? It does need to resolve conflicts with master.

shrenikm commented:
The quaternion base fixed point solver still needs to be fixed. I'll request a review once it's done and all conflicts have been resolved.

shrenikm commented:
## Updates

* Added quaternion floating base tests to `multibody_solvers_test`, `contact_toolkit_test` and `constrainted_lqr_controller_test`.
* Fixed Eigen bugs with the quaternion implementation of the solvers and controller.
* Made the changes requested in the previous review.
* Moved the `StateConnector` (Utility connector system that produces an `OutputVector` output given a `BasicVector` input of the state) to `multibody/utility_systems`.
* Updated the controller computations to fix bugs caused due to the quaternion models. Two separate Jacobians are computed now.
* `ContactToolkit` now has a `in_terms_of_qdot` boolean member that is used to compute the type of Jacobian. By default, if the number of positions and velocities are equal (Ex: RPY base), `in_terms_of_qdot=true`, else (Ex: Quaternion base) `in_terms_of_qdot=false`.
* `CalcContactJacobian` in `ContactToolkit` now is an overloaded function that can take in a `in_terms_of_qdot` boolean argument and compute the Jacobian accordingly. This is useful in the case where the `ContactToolkit` object is created using a specific tree, and both types of Jacobians are required to be computed from it. (See `systems/controllers/constrained_lqr_controller`).
* The controller simulation has been tested (Slow with contact).
* The added functionality and tests have been documented.

mposa commented on systems/controllers/affine_controller.cc:
```diff
@@ -7,35 +7,38 @@ using Eigen::VectorXd;
       this->EvalVectorInput(context, input_port_params_index_));
 
-  // could use MatrixXd instead of auto, but I think that would force a copy
+  // Using the Eigen type instead of auto would force a copy.
```
`auto` is still generating an Eigen type, but it's a Eigen `Map`, which acts as a reference. I think the original comment was more accurate.

mposa commented on systems/controllers/affine_controller.h:
```diff
@@ -58,45 +61,49 @@ class AffineController : public drake::systems::LeafSystem<double> {
-                     num_states) {}
-
-  int GetNumStates() const { return num_states_;}
+  AffineParams(int num_states, int num_efforts)
```
Is this spacing really better?

shrenikm replied:
This (And the other formatting differences) is the output of running my clang formatter on the file. Not really sure what to do about it.

shrenikm commented on attic/multibody/multibody_solvers.cc:
```diff
@@ -614,7 +646,7 @@ void FixedPointSolver::AddJointLimitConstraint(const double tolerance) {
   VectorXd joint_max = tree_.joint_limit_max;
 
-  for (int i = 0; i < joint_min.size(); ++i) {
+  for (int i = 7; i < joint_min.size(); ++i) {
```
I've updated it to ignore the positions with names starting with "base". It isn't the most effective way of doing it but is probably better than the hardcoded version.
